src/io/binary_parser.cpp

```diff
@@ -276,3 +289,36 @@ long long read_edges_inc(string filename, dd::FactorGraph &fg) {
   file.close();
   return count;
 }
+
+// read domains for multinomial
+void read_domains(std::string filename, dd::FactorGraph &fg) {
+  ifstream file;
+  file.open(filename.c_str(), ios::in | ios::binary);
+  long id, value;
+  long cardinality;
+  while (true) {
+    file.read((char *)&id, 8);
+    if (!file.good()) break;
+
+    id = be64toh(id);
+    dd::Variable &variable = fg.variables[id];
+    cardinality = variable.upper_bound - variable.lower_bound + 1;
```

Comment:
Can we get rid of `upper_bound` and `lower_bound`, and just keep one `cardinality` which can be used instead of `domain.size()` as well?